### ISSUE TITLE
Minor queue crash when upgrading from 3.9.14 to 3.10.rc.1

### ISSUE THREAD

**Issue by johanrhodin** [rabbitmq/rabbitmq-server#4373]
When upgrading from `3.9.14` to `3.10.rc-1` a queue crash is logged. It seems benign (no messages are lost), but contains thousands of log lines.

How to reproduce:
(1 node running 3.9.14)

Load up a lot of messages (I don't think the 10M really matters, but it was what I used) to a Durable classic queue without any other policy.
```
bin/runjava com.rabbitmq.perf.PerfTest -y0 -p -u "smallish" \
  -s 0 --id "smallish" -ad false -f persistent \
 -h $TARGET  \
 -pmessages 10000000
```
Upgrade to `3.10.rc.1` (`dpkg -i rabbitmq-server_3.10.0.rc.1-1_all.deb`)
The attached log is generated.
Some key parts:

```2022-03-29 16:14:46.216854+00:00 [erro] <0.463.0>   crasher:
2022-03-29 16:14:46.216854+00:00 [erro] <0.463.0>     initial call: rabbit_prequeue:init/1
2022-03-29 16:14:46.216854+00:00 [erro] <0.463.0>     pid: <0.463.0>
2022-03-29 16:14:46.216854+00:00 [erro] <0.463.0>     registered_name: []
2022-03-29 16:14:46.216854+00:00 [erro] <0.463.0>     exception exit: {function_clause,
2022-03-29 16:14:46.216854+00:00 [erro] <0.463.0>                         [{rabbit_variable_queue,d,
2022-03-29 16:14:46.216854+00:00 [erro] <0.463.0>                              [{delta,0,9999894,0,0}],
2022-03-29 16:14:46.216854+00:00 [erro] <0.463.0>                              [{file,"rabbit_variable_queue.erl"},{line,1587}]},
2022-03-29 16:14:46.216854+00:00 [erro] <0.463.0>                          {rabbit_variable_queue,init,11,
2022-03-29 16:14:46.216854+00:00 [erro] <0.463.0>                              [{file,"rabbit_variable_queue.erl"},{line,1802}]},
2022-03-29 16:14:46.216854+00:00 [erro] <0.463.0>                          {rabbit_priority_queue,init,3,
2022-03-29 16:14:46.216854+00:00 [erro] <0.463.0>                              [{file,"rabbit_priority_queue.erl"},{line,150}]},
2022-03-29 16:14:46.216854+00:00 [erro] <0.463.0>                          {rabbit_amqqueue_process,init_it2,3,
2022-03-29 16:14:46.216854+00:00 [erro] <0.463.0>                              [{file,"rabbit_amqqueue_process.erl"},
2022-03-29 16:14:46.216854+00:00 [erro] <0.463.0>                               {line,209}]},
2022-03-29 16:14:46.216854+00:00 [erro] <0.463.0>                          {rabbit_amqqueue_process,handle_call,3,
2022-03-29 16:14:46.216854+00:00 [erro] <0.463.0>                              [{file,"rabbit_amqqueue_process.erl"},
2022-03-29 16:14:46.216854+00:00 [erro] <0.463.0>                               {line,1294}]},
2022-03-29 16:14:46.216854+00:00 [erro] <0.463.0>                          {gen_server2,handle_msg,2,
2022-03-29 16:14:46.216854+00:00 [erro] <0.463.0>                              [{file,"gen_server2.erl"},{line,1046}]},
2022-03-29 16:14:46.216854+00:00 [erro] <0.463.0>                          {proc_lib,init_p_do_apply,3,
2022-03-29 16:14:46.216854+00:00 [erro] <0.463.0>                              [{file,"proc_lib.erl"},{line,226}]}]}
2022-03-29 16:14:46.216854+00:00 [erro] <0.463.0>       in function  gen_server2:terminate/3 (gen_server2.erl, line 1183)
2022-03-29 16:14:46.216854+00:00 [erro] <0.463.0>     ancestors: [<0.462.0>,<0.460.0>,<0.445.0>,<0.444.0>,
2022-03-29 16:14:46.216854+00:00 [erro] <0.463.0>                   rabbit_vhost_sup_sup,rabbit_sup,<0.224.0>]
2022-03-29 16:14:46.216854+00:00 [erro] <0.463.0>     message_queue_len: 0
2022-03-29 16:14:46.216854+00:00 [erro] <0.463.0>     messages: []
2022-03-29 16:14:46.216854+00:00 [erro] <0.463.0>     links: [<0.462.0>]
```
and later:
`Restarting crashed queue 'smallish' in vhost 'werflyha'.`

The next time RabbitMQ restarts the queue will be marked as "dirty" (not sure that is the correct word) and will have to be rebuilt from scratch. (`recovery terms differ from present`-> `rebuilding indices from scratch`). Restarting RabbitMQ again after that goes fine.

[queue_crash.log](https://github.com/rabbitmq/rabbitmq-server/files/8373447/queue_crash.log)

**Comment by lhoguin** [comment]
Hello and thanks for the report! I think the patch to fix this issue is:

``` diff
diff --git a/deps/rabbit/src/rabbit_variable_queue.erl b/deps/rabbit/src/rabbit_variable_queue.erl
index 34e9ca4870..d988fc0ec3 100644
--- a/deps/rabbit/src/rabbit_variable_queue.erl
+++ b/deps/rabbit/src/rabbit_variable_queue.erl
@@ -1788,7 +1788,7 @@ init(QueueVsn, IsDurable, IndexMod, IndexState, StoreState, DeltaCount, DeltaByt
         case Terms of
             non_clean_shutdown -> {HiSeqId, HiSeqId, DeltaCount, DeltaBytes};
             _                  -> NextSeqId0 = proplists:get_value(next_seq_id,
-                                                                   Terms, 0),
+                                                                   Terms, HiSeqId),
                                   {NextSeqId0,
                                    proplists:get_value(next_deliver_seq_id,
                                                        Terms, NextSeqId0),
```

Although I have not tested it yet.

**Comment by johanrhodin** [comment]
Nice job 🙌 
Testing it would involve checking out the branch, build it (.deb), and re-install? I'm unfortunately not able to do that today.

**Comment by mkuratczyk** [comment]
I'm afraid we don't build all the packages for every commit/PR so if you want to use .deb files - yes. However, we do most of the tests with OCI images that we do build for every commit/PR. You can find this version here: https://hub.docker.com/layers/rabbitmq/pivotalrabbitmq/rabbitmq/loic-fix-cq-upgrade-3.10-otp-max/images/sha256-eff62e6a25c957460f086a202415072aa11e761f1ea235e81abb1b22de5be183

And in general, you can use `pivotalrabbitmq/rabbitmq/BRANCH_NAME-otp-max` (or "otp-min" for a build with the oldest support Erlang version) to get a Docker image with the most recent commits on a given branch (on Kubernetes, use `imagePullPolicy: Always` to make sure you actually pull with the latest changes, in case commits are added to the branch.

**Comment by michaelklishin** [comment]
This will ship in `3.10.0-rc.2` hopefully next week.
